pg_sparse/src/index_access/scan.rs

```diff
@@ -0,0 +1,138 @@
+    let mut scandesc: PgBox<pg_sys::IndexScanDescData> =
+        unsafe { PgBox::from_pg(pg_sys::RelationGetIndexScan(indexrel, nkeys, norderbys)) };
+    let index_relation = unsafe { PgRelation::from_pg(indexrel) };
+    let index_name = index_relation.name().to_string();
+    let index = from_index_name(&index_name);
+    let rdopts = get_rdopts(index_relation);
+
+    // Create the index and scan
+    let scan_state = ScanState {
+        index,
+        results: vec![],
+        current: 0,
+        n_results: 0,
+        no_more_results: false,
+        k: rdopts.ef_search as usize,
+        ef_search: rdopts.ef_search as usize,
+        query_vector: Sparse {
+            entries: vec![],
+            n: 0,
+        },
+    };
+
+    scandesc.opaque =
+        PgMemoryContexts::CurrentMemoryContext.leak_and_drop_on_delete(scan_state) as void_mut_ptr;
+    scandesc.into_pg()
+}
+
+#[pg_guard]
+pub extern "C" fn amrescan(
+    scan: pg_sys::IndexScanDesc,
+    _keys: pg_sys::ScanKey,
+    _nkeys: ::std::os::raw::c_int,
+    orderbys: pg_sys::ScanKey,
+    norderbys: ::std::os::raw::c_int,
+) {
+    // Convert the raw pointer to a safe wrapper. This action takes ownership of the object
+    // pointed to by the raw pointer in a safe way.
+    let scan: PgBox<pg_sys::IndexScanDescData> = unsafe { PgBox::from_pg(scan) };
+
+    // Extract the scan state from the opaque field of the scan descriptor.
+    let state = unsafe { (scan.opaque as *mut ScanState).as_mut() }.expect("No scandesc state");
+
+    state.results.clear();
+    state.current = 0;
+
+    // Retrieve the query vector
+    if !orderbys.is_null() && norderbys > 0 {
+        let orderbys_slice = unsafe { std::slice::from_raw_parts(orderbys, norderbys as usize) };
+        let sk_argument: Option<Sparse> =
+            unsafe { FromDatum::from_datum(orderbys_slice[0].sk_argument, false) };
+        state.query_vector = sk_argument.expect("Could not parse query vector");
+    }
+}
+
+#[pg_guard]
+pub extern "C" fn amendscan(_scan: pg_sys::IndexScanDesc) {}
+
+#[pg_guard]
+pub extern "C" fn amgettuple(
+    scan: pg_sys::IndexScanDesc,
+    direction: pg_sys::ScanDirection,
+) -> bool {
+    assert!(direction == pg_sys::ScanDirection_ForwardScanDirection);
+
+    // Extract the scan state from the opaque field of the scan descriptor.
+    let mut scan: PgBox<pg_sys::IndexScanDescData> = unsafe { PgBox::from_pg(scan) };
+    let state = unsafe { (scan.opaque as *mut ScanState).as_mut() }.expect("No scandesc state");
+
+    // First scan
+    if state.current == 0 {
+        let results =
+            state
+                .index
+                .search_knn(state.query_vector.entries.clone(), state.k, state.ef_search);
+
+        state.results = results.clone();
+        state.n_results = results.len();
+        state.no_more_results = state.n_results < state.k;
+    }
+
+    // Subsequent scans with larger k if necessary
+    if state.current >= state.n_results {
```

Comment:
Not sure I understand this logic. What is `state.current` and when is it incremented? Perhaps add a comment to clarify what this is doing.

Reply:
Sure, I can add a comment. Under the hood, Postgres calls amgettuple repeatedly until the function returns false (no more results found). State.current starts at 0 and increments by 1 on every call.